File: src/cmds/build_cmds.rs
//! Commands for building zrc toolchains

use std::{error::Error, process::Command};

use clap::Parser;

use crate::{cli::DispatchCommand, deps, git_utils, paths};

/// Build a specific version of zrc
#[derive(Parser)]
pub struct BuildCmd {
    /// The git reference to build (branch, tag, or commit)
    pub reference: String,

    /// Custom zrc repository URL
    #[arg(
        long = "zrc-repo",
        default_value = "https://github.com/zirco-lang/zrc.git"
    )]
    pub repo_url: String,
}

impl DispatchCommand for BuildCmd {
    fn dispatch(self) -> Result<(), Box<dyn Error>> {
        // Check dependencies first
        deps::check_dependencies_strict()?;

        // Ensure directories exist
        paths::ensure_directories()?;

        let source_dir = paths::zrc_source_dir();

        // Clone or open repository
        let repo = git_utils::clone_or_open(&self.repo_url, &source_dir)?;

        // Fetch latest changes
        git_utils::fetch(&repo)?;

        // Checkout the requested reference
        git_utils::checkout_ref(&repo, &self.reference)?;

        // Get commit SHA for version naming
        let commit_sha = git_utils::get_current_commit_short(&repo)?;

        // Determine reference type and create appropriate version name
        let ref_type = git_utils::determine_ref_type(&repo, &self.reference);
        let version = match ref_type {
            git_utils::RefType::Tag(tag) => tag,
            git_utils::RefType::Branch(branch) => {
                format!("{}@{}", branch.replace('/', "-"), commit_sha)
            }
            git_utils::RefType::Commit(commit) => commit, // No prefix for commits
        };

        println!("Building version: {}", version);

        // Create toolchain directory
        let toolchain_dir = paths::toolchain_dir(&version);
        std::fs::create_dir_all(&toolchain_dir)?;

        // Execute the hook script from the zrc repo
        // The hook handles building and installing to the toolchain directory
        run_build_hook(&source_dir, &toolchain_dir)?;

        // Update current symlink
        let current_link = paths::current_toolchain_link();
        paths::create_link(&toolchain_dir, &current_link)?;

        println!("\n✓ Successfully built and installed zrc {}", version);
        println!("  Toolchain location: {}", toolchain_dir.display());
        println!("\nTo use zrc, run:");
        println!("  source <(zircon env)");

        Ok(())
    }
}

/// Run the build hook script from the zrc repository
#[cfg(unix)]
fn run_build_hook(
    source_dir: &std::path::Path,
    toolchain_dir: &std::path::Path,
) -> Result<(), Box<dyn Error>> {
    let hook_script = source_dir.join("hooks").join("zircon.sh");
    if !hook_script.exists() {
        return Err(format!(
            "Hook script not found at {}. This version of zrc may not support zircon hooks.",
            hook_script.display()
        )
        .into());
    }

    println!("Running zrc build hook...");
    let status = Command::new("bash")
        .arg(&hook_script)
        .env("ZIRCON_TOOLCHAIN_DIR", toolchain_dir)
        .current_dir(source_dir)
        .status()?;

    if !status.success() {
        let exit_code = status.code().unwrap_or(-1);
        return Err(format!("Hook script failed (exit code: {})", exit_code).into());
    }

    Ok(())
}

/// Run the build hook script from the zrc repository (Windows)
#[cfg(windows)]
fn run_build_hook(
    source_dir: &std::path::Path,
    toolchain_dir: &std::path::Path,
) -> Result<(), Box<dyn Error>> {
    // Check for PowerShell script first, then batch file
    let ps_hook = source_dir.join("hooks").join("zircon.ps1");
    let bat_hook = source_dir.join("hooks").join("zircon.bat");

    if ps_hook.exists() {
        println!("Running zrc build hook (PowerShell)...");
        // Use Bypass to run local scripts regardless of system execution policy.
        // This is safe because the script is part of the zrc repo the user cloned.
        let status = Command::new("powershell")
            .args(["-ExecutionPolicy", "Bypass", "-File"])
            .arg(&ps_hook)
            .env("ZIRCON_TOOLCHAIN_DIR", toolchain_dir)
            .current_dir(source_dir)
            .status()?;

        if !status.success() {
            let exit_code = status.code().unwrap_or(-1);
            return Err(format!("Hook script failed (exit code: {})", exit_code).into());
        }
    } else if bat_hook.exists() {
        println!("Running zrc build hook (batch)...");
        let status = Command::new("cmd")
            .args(["/C"])
            .arg(&bat_hook)
            .env("ZIRCON_TOOLCHAIN_DIR", toolchain_dir)
            .current_dir(source_dir)
            .status()?;

        if !status.success() {
            let exit_code = status.code().unwrap_or(-1);
            return Err(format!("Hook script failed (exit code: {})", exit_code).into());
        }
    } else {
        return Err(format!(
            "No Windows hook script found at {} or {}. \
             This version of zrc may not support Windows builds via zircon hooks.",
            ps_hook.display(),
            bat_hook.display()
        )
        .into());
    }

    Ok(())
}

File: src/cmds/build_cmds.c
#include "build_cmds.h"
#include "../deps.h"
#include "../git_utils.h"
#include "../paths.h"

static int run_build_hook(const char* source_dir, const char* toolchain_dir) {
    char hook_path[MAX_PATH];
    snprintf(hook_path, sizeof(hook_path), "%s/hooks/zircon.sh", source_dir);
    
    if (!path_exists(hook_path)) {
        fprintf(stderr, "Hook script not found at %s.\n", hook_path);
        fprintf(stderr, "This version of zrc may not support zircon hooks.\n");
        return -1;
    }
    
    printf("Running zrc build hook...\n");
    
    char cmd[MAX_CMD];
    snprintf(cmd, sizeof(cmd), 
             "cd '%s' && ZIRCON_TOOLCHAIN_DIR='%s' bash '%s'",
             source_dir, toolchain_dir, hook_path);
    
    int status = system(cmd);
    if (status != 0) {
        fprintf(stderr, "Hook script failed (exit code: %d)\n", status);
        return -1;
    }
    
    return 0;
}

int dispatch_build_command(cli_context_t* ctx) {
    /* Check dependencies first */
    if (check_dependencies_strict() != 0) {
        return ERR_GENERAL;
    }
    
    /* Ensure directories exist */
    if (ensure_directories() != 0) {
        fprintf(stderr, "Failed to create directories\n");
        return ERR_IO;
    }
    
    char* source_dir = zrc_source_dir();
    if (!source_dir) return ERR_GENERAL;
    
    /* Clone or open repository */
    git_repository* repo = NULL;
    if (clone_or_open(ctx->data.build_cmd.repo_url, source_dir, &repo) != 0) {
        fprintf(stderr, "Failed to clone/open repository\n");
        free(source_dir);
        return ERR_GIT;
    }
    
    /* Fetch latest changes */
    if (fetch_repo(repo) != 0) {
        fprintf(stderr, "Failed to fetch repository\n");
        git_repository_free(repo);
        free(source_dir);
        return ERR_GIT;
    }
    
    /* Checkout the requested reference */
    if (checkout_ref(repo, ctx->data.build_cmd.reference) != 0) {
        fprintf(stderr, "Failed to checkout reference: %s\n", ctx->data.build_cmd.reference);
        git_repository_free(repo);
        free(source_dir);
        return ERR_GIT;
    }
    
    /* Get commit SHA for version naming */
    char* commit_sha = get_current_commit_short(repo);
    if (!commit_sha) {
        fprintf(stderr, "Failed to get commit SHA\n");
        git_repository_free(repo);
        free(source_dir);
        return ERR_GIT;
    }
    
    /* Determine reference type and create appropriate version name */
    ref_info_t* ref_info = determine_ref_type(repo, ctx->data.build_cmd.reference);
    if (!ref_info) {
        fprintf(stderr, "Failed to determine reference type\n");
        free(commit_sha);
        git_repository_free(repo);
        free(source_dir);
        return ERR_GIT;
    }
    
    char version[MAX_PATH];
    if (ref_info->type == REF_TYPE_TAG) {
        snprintf(version, sizeof(version), "%s", ref_info->name);
    } else if (ref_info->type == REF_TYPE_BRANCH) {
        char* safe_branch = string_replace_char(ref_info->name, '/', '-');
        snprintf(version, sizeof(version), "%s@%s", safe_branch, commit_sha);
        free(safe_branch);
    } else {
        snprintf(version, sizeof(version), "%s", ref_info->name);
    }
    
    printf("Building version: %s\n", version);
    
    free_ref_info(ref_info);
    free(commit_sha);
    git_repository_free(repo);
    
    /* Create toolchain directory */
    char* tc_dir = toolchain_dir(version);
    if (!tc_dir) {
        free(source_dir);
        return ERR_GENERAL;
    }
    
    if (make_directory_recursive(tc_dir) != 0) {
        fprintf(stderr, "Failed to create toolchain directory\n");
        free(tc_dir);
        free(source_dir);
        return ERR_IO;
    }
    
    /* Execute the hook script */
    if (run_build_hook(source_dir, tc_dir) != 0) {
        free(tc_dir);
        free(source_dir);
        return ERR_GENERAL;
    }
    
    /* Update current symlink */
    char* current_link = current_toolchain_link();
    if (current_link) {
        create_link(tc_dir, current_link);
        free(current_link);
    }
    
    printf("\n✓ Successfully built and installed zrc %s\n", version);
    printf("  Toolchain location: %s\n", tc_dir);
    printf("\nTo use zrc, run:\n");
    printf("  source <(zircon env)\n");
    
    free(tc_dir);
    free(source_dir);
    
    return ERR_OK;
}
